iceoryx_utils/include/iceoryx_utils/internal/units/duration.inl

```diff
@@ -20,262 +20,511 @@ namespace iox
+    : Duration(static_cast<Seconds_t>(value.tv_sec), static_cast<Nanoseconds_t>(value.tv_usec) * NANOSECS_PER_MICROSEC)
 {
 }
 
-inline constexpr Duration::Duration(const struct timespec& value)
-    : Duration(static_cast<long double>(value.tv_sec) + static_cast<long double>(value.tv_nsec) / 1000000000.0)
+inline constexpr Duration::Duration(const struct timespec& value) noexcept
+    : Duration(static_cast<Seconds_t>(value.tv_sec), static_cast<Nanoseconds_t>(value.tv_nsec))
 {
 }
 
-inline constexpr Duration::Duration(const struct itimerspec& value)
+inline constexpr Duration::Duration(const struct itimerspec& value) noexcept
     : Duration(value.it_interval)
 {
 }
 
-inline constexpr Duration::Duration(const std::chrono::milliseconds& value)
-    : Duration(static_cast<long double>(value.count()) / 1000.0)
+inline constexpr Duration::Duration(const std::chrono::milliseconds& value) noexcept
 {
+    *this = Duration::milliseconds(value.count());
 }
 
-inline constexpr Duration::Duration(const std::chrono::nanoseconds& value)
-    : Duration(static_cast<long double>(value.count()) / 1000000000.0)
+inline constexpr Duration::Duration(const std::chrono::nanoseconds& value) noexcept
 {
+    *this = Duration::nanoseconds(value.count());
 }
 
-inline Duration& Duration::operator=(const std::chrono::milliseconds& right)
+inline Duration& Duration::operator=(const std::chrono::milliseconds& rhs) noexcept
 {
-    durationInSeconds = static_cast<long double>(right.count()) / 1000.0;
+    *this = Duration(rhs);
     return *this;
 }
 
-inline constexpr Duration::Duration(const long double durationInSeconds)
-    : durationInSeconds(durationInSeconds >= 0.0 ? durationInSeconds : 0.0)
+inline constexpr uint64_t Duration::nanoSeconds() const noexcept
 {
-}
+    constexpr Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Seconds_t>::max() / NANOSECS_PER_SEC};
+    constexpr Nanoseconds_t MAX_NANOSECONDS_BEFORE_OVERFLOW{std::numeric_limits<Seconds_t>::max() % NANOSECS_PER_SEC};
+    constexpr Duration MAX_DURATION_BEFORE_OVERFLOW =
+        createDuration(MAX_SECONDS_BEFORE_OVERFLOW, MAX_NANOSECONDS_BEFORE_OVERFLOW);
 
-template <typename T>
-inline constexpr T Duration::nanoSeconds() const
-{
-    return static_cast<T>(durationInSeconds * 1000000000.0);
+    if (*this > MAX_DURATION_BEFORE_OVERFLOW)
+    {
+        std::clog << __PRETTY_FUNCTION__ << ": Result of conversion would overflow, clamping to max value!"
+                  << std::endl;
+        return std::numeric_limits<uint64_t>::max();
+    }
+
+    return m_seconds * NANOSECS_PER_SEC + m_nanoseconds;
 }
 
-template <typename T>
-inline constexpr T Duration::microSeconds() const
+inline constexpr uint64_t Duration::microSeconds() const noexcept
 {
-    return static_cast<T>(durationInSeconds * 1000000.0);
+    constexpr Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Seconds_t>::max() / MICROSECS_PER_SEC};
```

Comment:
I am not sure the overflow duration is correct in general in this way, since it actually depends on the return type (uint64_t). If this is a type larger than Seconds_t the boundary would be to pessimistic I think. Returning max of uint64 is also generally not correct in those cases (it might actually fit into uint64_t and be less).

Currently it is fine since Seconds_t is uint64_t.

I would not think this is a bug for this reason. Moving on.

Reply:
you are right, it should be the return type